File: 9-Stack/programs/ruby/program1.rb
# Stack implementation

class Stack
  def initialize(size)
    @size = size
    @top = -1
    @stack = Array.new(size)
  end

  def push(x)
    if @top == @size - 1
      puts "Stack Overflow"
    else
      @top += 1
      @stack[@top] = x
    end
  end

  def pop
    if @top == -1
      puts "Stack Underflow"
      -1
    else
      x = @stack[@top]
      @top -= 1
      x
    end
  end

  def peek(index)
    if @top - index + 1 < 0
      puts "Invalid Index"
      -1
    else
      @stack[@top - index + 1]
    end
  end

  def stack_top
    @top == -1 ? -1 : @stack[@top]
  end

  def is_empty
    @top == -1
  end

  def is_full
    @top == @size - 1
  end

  def display
    for i in (0..@top).reverse_each
      print "#{@stack[i]} "
    end
    puts
  end
end

stack = Stack.new(5)
stack.push(10)
stack.push(20)
stack.push(30)
stack.push(40)
stack.push(50)
stack.push(60)

puts "Popped: #{stack.pop}"
puts "Peek: #{stack.peek(2)}"
stack.display

File: 9-Stack/programs/cpp/program1.cpp
// Stack implementation

#include<bits/stdc++.h>
using namespace std;

struct Stack{
  int size;
  int top;
  int *S;
};

void create(struct Stack *st){
  cout << "Enter size: ";
  cin >> st->size;
  cout << st->size << endl;
  st->top = -1;
  st->S = new int[st->size];
}

void Display(struct Stack st){
  for (int i = st.top; i >= 0; i--)
  {
    cout << st.S[i] << " ";
  }
  cout << endl;
}

void push(struct Stack *st, int x){
  if(st->top == st->size-1){
    cout << "Stack Overflow" << endl;
  }else{
    st->top++;
    st->S[st->top] = x;
  }
}

int pop(struct Stack *st){
  int x = -1;
  if(st->top == -1){
    cout << "Stack Underflow" << endl;
  }else{
    x = st->S[st->top];
    st->top--;
  }
  return x;
}

int peek(struct Stack st, int index){
  int x = -1;
  if(st.top-index+1 < 0){
    cout << "Invalid Index" << endl;
  }else{
    x = st.S[st.top-index+1];
  }
  return x;
}

int stackTop(struct Stack st){
  if(st.top == -1){
    return -1;
  }else{
    return st.S[st.top];
  }
}

int isEmpty(struct Stack st){
  return st.top == -1;
}

int isFull(struct Stack st){
  return st.top == st.size-1;
}

int main(){
  struct Stack st;
  create(&st);

  push(&st, 10);
  push(&st, 20);
  push(&st, 30);
  push(&st, 40);
  push(&st, 50);
  push(&st, 60);

  cout << "Popped: " << pop(&st) << endl;

  cout << "Peek: " << peek(st, 2) << endl;

  Display(st);

  return 0;
}
